serializer_impl.hh

```diff
@@ -782,4 +782,138 @@ unknown_variant_type deserialize(Input& in, boost::type<unknown_variant_type>) {
         return unknown_variant_type{ index, std::move(v) };
     });
 }
+
+template <typename T>
+class range_deserializer {
+    utils::input_stream _in;
+    uint32_t _size;
+
+public:
+    range_deserializer(const utils::input_stream& in)
+        : _in(in)
+    {
+        _size = deserialize(_in, boost::type<uint32_t>());
+    }
+
+    // Get the number of items in the vector
+    size_t size() const noexcept {
+        return _size;
+    }
+
+    class iterator;
+
+    // Sentinel for the end() position
+    class sentinel {
+        uint32_t _idx;
+
+        sentinel(uint32_t idx) noexcept
+            : _idx(idx)
+        { }
+
+        friend class range_deserializer;
+    public:
+        ssize_t operator-(const iterator& it) const noexcept {
+            return _idx - it._idx;
+        }
+    };
+
+    // Forward iterator
+    class iterator {
+        utils::input_stream _in;
+        uint32_t _idx = 0;
+        std::optional<T> _item;
+
+        iterator(const utils::input_stream& in, uint32_t idx, std::optional<T>&& item)
+            : _in(in)
+            , _idx(idx)
+            , _item(std::move(item))
+        { }
+
+        friend class range_deserializer;
+    protected:
+        iterator(const utils::input_stream& in)
+            : _in(in)
+            , _idx(0)
+        { }
+
+    private:
+        static size_type read_frame_size(utils::input_stream& in) {
+            auto sz = deserialize(in, boost::type<size_type>());
+            if (sz < sizeof(size_type)) {
+                throw std::runtime_error(fmt::format("IDL frame truncated: expected to have at least {} bytes, got {}", sizeof(size_type), sz));
+            }
+            return sz - sizeof(size_type);
+        }
+
+   public:
+        iterator() = default;
+
+        iterator(iterator&& o) noexcept
+            : _in(std::move(o._in))
+            , _idx(o._idx)
+            , _item(std::exchange(o._item, std::nullopt))
+        { }
```

Comment:
Isn't = default (or rather, not defining anything) sufficient? One isn't supposed to use a forward iterator to read the same thing twice, and you aren't preventing it, since the copy constructor and assignment operators are still there.

Reply:
The problem with the default move constructor is that std::optional move constructor doesn't
make the other optional empty.
https://en.cppreference.com/w/cpp/utility/optional/optional

Reply:
Why is that a problem? I don't think we have any requirements w.r.t. dereferencing moved-from iterators.

Reply:
That's a good point.  I think it can indeed be removed.